api/metric/api.go

```diff
@@ -21,62 +23,231 @@ import (
 )
 
-type MetricType int
+// Kind categorizes different kinds of metric.
+type Kind int
 
 const (
-	Invalid    MetricType = iota
-	Gauge                 // Supports Set()
-	Cumulative            // Supports Inc()
+	Invalid     Kind = iota
+	CounterKind      // Supports Add()
+	GaugeKind        // Supports Set()
+	MeasureKind      // Supports Record()
+	ObserverKind
 )
 
-type Meter interface {
-	// TODO more Metric types
-	GetFloat64Gauge(ctx context.Context, gauge *Float64GaugeHandle, labels ...core.KeyValue) Float64Gauge
+// Recorder is the implementation-level interface to Set/Add/Record individual metrics.
+type Handle interface {
+	// Record allows the SDK to observe a single metric event
+	RecordOne(ctx context.Context, value MeasurementValue)
 }
 
-type Float64Gauge interface {
-	Set(ctx context.Context, value float64, labels ...core.KeyValue)
+// LabelSet represents a []core.KeyValue for use as pre-defined labels
+// in the metrics API.
+//
+// TODO this belongs outside the metrics API, in some sense, but that
+// might create a dependency.  Putting this here means we can't re-use
+// a LabelSet between metrics and tracing, even when they are the same
+// SDK.
+type LabelSet interface {
+	Meter() Meter
 }
 
-type Handle struct {
-	Name        string
-	Description string
-	Unit        unit.Unit
+// ObserverCallback defines a type of the callback SDK will call for
+// the registered observers.
+type ObserverCallback func(Meter, Observer) (LabelSet, MeasurementValue)
+
+// Meter is an interface to the metrics portion of the OpenTelemetry SDK.
+type Meter interface {
+	// DefineLabels returns a reference to a set of labels that
+	// cannot be read by the application.
+	DefineLabels(context.Context, ...core.KeyValue) LabelSet
+
+	NewHandle(context.Context, Descriptor, LabelSet) Handle
+	DeleteHandle(Handle)
+
+	// RecordBatch atomically records a batch of measurements..
+	RecordBatch(context.Context, LabelSet, ...Measurement)
+
+	RegisterObserver(Observer, ObserverCallback)
+	UnregisterObserver(Observer)
+}
+
+type DescriptorID uint64
+type MetricValueKind int8
+
+const (
+	AnyValueKind MetricValueKind = iota
+	Int64ValueKind
+	Float64ValueKind
+)
+
+// Descriptor represents a named metric with recommended local-aggregation keys.
+type Descriptor struct {
+	// Name is a required field describing this metric descriptor,
+	// should have length > 0.
+	Name string
 
-	Type MetricType
+	// Kind is the metric kind of this descriptor.
+	Kind Kind
+
+	// Keys are required keys determined in the handles
+	// obtained for this metric.
 	Keys []core.Key
+
+	// ID is uniquely assigned to support per-SDK registration.
+	ID DescriptorID
+
+	// Description is an optional field describing this metric descriptor.
+	Description string
+
+	// Unit is an optional field describing this metric descriptor.
+	Unit unit.Unit
+
+	// ValueKind describes the type of values the metric produces.
+	ValueKind MetricValueKind
+
+	// Meaning depends on the metric type:
```

Comment:
Rephrase "Meaning depends on the metric type:" to
"Alternate defines the property of metric value depend on metric type."
Also define what the value means. Is it Alternate == false mean the Counter can go up?

Reply:
Thanks for the suggestion.

The Counter can always go up. ;) But reworded it like "for Counter, true implies that the metric is an up-down Counter". I think that the docs for each metric will mention the default expected behaviour for recorded values (counter and measure take only non-negative values, gauge takes values greater than old ones).

Reply:
why gauge takes values greater than old ones? It can be lower than the old one.

Reply:
Eek, sorry. My mistake. Gauge can take lower values than old ones by default. The alternate version of gauge is non-descending one.